### DISCUSSION TITLE
How to achieve this effect with the widget.NewTable?

### DISCUSSION THREAD

**Discussion by junmaqiang** [fyne-io/fyne#3556] [Q&A] [answered]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

![WX20230111-143510](https://user-images.githubusercontent.com/1819457/211735197-aeb1a4c5-46fc-450f-b3fd-3b1899333243.png)

### Describe the solution you'd like to see.

How to achieve this effect with the widget.NewTable?

```
	templ := canvas.NewRectangle(color.Gray16{})
	templ.SetMinSize(fyne.Size{Width: width, Height: height})

	table := NewTable(
		func() (int, int) { return maxRows, maxCols },
		func() fyne.CanvasObject { return templ },
		func(TableCellID, fyne.CanvasObject) {})

	w := test.NewWindow(table)
```

my diy code
```
	th.CreateCell = func() fyne.CanvasObject {
		// my diy code
		bg := canvas.NewRectangle(color.Gray16{})
		bg.SetMinSize(fyne.Size{Width: 50, Height: 50})
		tv := widget.NewLabel("Devices")
		tb := container.NewWithoutLayout(bg, tv)
		return tb
	}
	th.UpdateCell = func(id widget.TableCellID, cell fyne.CanvasObject) {
		// fmt.Printf("%#v", cell.(*fyne.Container[(*widget.Label)])) // How to call custom component content?
		// label := cell.(*widget.Label) // This is the official way of writing
        }
```

**Comment by Jacalz** [comment]
Please use GitHub discussions for questions. We prefer to keep the issues for bugs and feature requests.

I converted this into a discussion.

**Comment by Jacalz** [comment]
I believe you just want to use a `container.NewBorder()` and then have your buttons (the ones you have marked within a red rectangle) in a `container.NewHBox()` as the “right” parameter to the border.

Looking at your DIY code, I would avoid the rectangle and containers without layout entirely. They do no scale or update colour when things change. I think watching https://m.youtube.com/watch?v=LWn1403gY9E&embeds_euri=https%3A%2F%2Fdeveloper.fyne.io%2F&feature=emb_imp_woyt could be helpful for you.

**Comment by junmaqiang** [comment]
@Jacalz  thank you，  I would like to ask you how to realize the multiple expressions of column fields in the table?   There's text, there's buttons, there's hyperlinks, but I haven't seen anyone else write CreateCell here, right

**Reply by Jacalz** [reply]
You basically just want to combine containers and layouts in a way to get the desired output. As for how to use the functions, I suggest looking at the code in fyne_demo: https://github.com/fyne-io/fyne/blob/8c2509518b2df442a6b748d9b07754739592e6d7/cmd/fyne_demo/tutorials/collection.go#L64

**Reply by Jacalz** [reply]
You might also want to look at how the List widget is using the data[id] to use a baking data slice for updating contents: https://github.com/fyne-io/fyne/blob/8c2509518b2df442a6b748d9b07754739592e6d7/cmd/fyne_demo/tutorials/collection.go#L24
You can do something similar with Table.

**Reply by andydotxyz** [reply]
I wrote an article on this topic recently. The trick is to define all the different types of content and hide/show items according to cell id.

https://fynelabs.com/2022/12/30/building-complex-tables-with-fyne/

**Reply by Jacalz** [reply]
Ah, right. Looking at the image above, my mind was basically thinking in terms of the List widget. It would be easier to replicate the image above in that way, I think.

**Reply by junmaqiang** [reply]
thank you i know，just this i want

**Comment by junmaqiang** [comment, marked as answer]
@andydotxyz very very thank you  https://fynelabs.com/2022/12/30/building-complex-tables-with-fyne/ is good
